src/share/classes/com/sun/javatest/regtest/agent/TestNGRunner.java

```diff
@@ -109,10 +113,18 @@ public static class RegressionListener
             implements ITestListener, IConfigurationListener {
         enum InfoKind { CONFIG, TEST }
 
+        // keeps track of the test start time for each test
+        private final Map<ITestResult, Long> startTimeNanos =
+                Collections.synchronizedMap(new IdentityHashMap<>());
```

Comment:
Why didn't you use `java.util.concurrent.ConcurrentHashMap` here?

Reply:
Hello Christian, we want the equality check of the key to be based on identity (since I don't think we can rely on the `equals()` implementation of `ITestResult`). ConcurrentHashMap uses `equals()` check unlike `IdentityHashMap` which uses `==`.

In the case of junit, where you used a ConcurrentHashMap, i think it's fine since the tests in junit appear to have a `UniqueId` unlike here where I don't see that construct exposed in testng.

Reply:
Understood.

Meaning that in this case we rely on TestNG re-using the exact same object instances for `onTestStart()` and `report()` methods.

Reply:
A possibility here would be to maybe use a record class for the concurrent hash map key which extracts the test class name, and the test name from ITestResult. Would that be unique enough?

Reply:
Literally `record` types are too new for agent classes - only Java 8 is supported here.